File: laundry/wash_wallet_be/resources/js/types/order.ts
import {
    BaseFilters,
    BaseSortOptions,
    ReferralLog,
    Customer,
    Employee,
    OrderItem,
    OrderStatusHistory,
    Outlet,
} from ".";

export interface OrderReview {
    id: number;
    rating: number;
    review?: string | null;
    createdAt: string;
    formattedCreatedAt: string;
}

export interface CustomerAddress {
    id: number;
    label?: string | null;
    address: string;
    city?: string | null;
    province?: string | null;
    postalCode?: string | null;
    latitude?: number | null;
    longitude?: number | null;
}

export type OrderPaymentMethod = "cash" | "transfer" | "qris" | "debit" | "cod";

export interface OrderPaymentLog {
    id: number;
    orderId: number;
    employeeId?: number | null;
    amount: number;
    formattedAmount: string;
    paymentMethod: OrderPaymentMethod;
    paymentMethodLabel: string;
    referenceNumber?: string | null;
    notes?: string | null;
    employee?: Employee;
    createdAt: string;
    formattedCreatedAt: string;
    canBeDeleted: boolean;
}

export interface Order {
    id: number;
    orderNumber: string;
    source: string;
    sourceLabel: string;
    status:
        | "requested"
        | "cancelled"
        | "accepted"
        | "rejected"
        | "picking_up"
        | "received"
        | "weighing"
        | "ready_to_process"
        | "in_progress"
        | "ready"
        | "delivering"
        | "delivered"
        | "completed";
    statusLabel: string;
    statusBadgeVariant:
        | "default"
        | "primary"
        | "secondary"
        | "success"
        | "warning"
        | "info"
        | "error";
    paymentStatus:
        | "not_yet_priced"
        | "unpaid"
        | "partial"
        | "paid"
        | "refunded"
        | "paid_by_package"
        | "cod";
    paymentStatusLabel: string;
    paymentStatusBadgeVariant:
        | "default"
        | "primary"
        | "secondary"
        | "success"
        | "warning"
        | "info"
        | "error";
    deliveryType: string;
    deliveryTypeLabel: string;
    completionPercentage: number;
    customerId: number;
    customerAccountId?: number | null;
    employeeId: number;
    outletId: number;
    customerAddressId?: number | null;
    updatedBy?: number | null;
    subtotal: number;
    formattedSubtotal: string;
    discountAmount: number;
    formattedDiscountAmount: string;
    taxAmount: number;
    formattedTaxAmount: string;
    pickupFee: number;
    formattedPickupFee: string;
    deliveryFee: number;
    formattedDeliveryFee: string;
    totalAmount: number;
    formattedTotalAmount: string;
    paidAmount: number;
    formattedPaidAmount: string;
    remainingAmount: number;
    formattedRemainingAmount: string;
    orderDate: string;
    formattedOrderDate: string;
    estimatedCompletion?: string | null;
    formattedEstimatedCompletion?: string | null;
    actualCompletion?: string | null;
    formattedActualCompletion?: string | null;
    pickupDate?: string | null;
    formattedPickupDate?: string | null;
    pickupAddress?: string | null;
    pickupSchedule?: string | null;
    formattedPickupSchedule?: string | null;
    deliveryDate?: string | null;
    formattedDeliveryDate?: string | null;
    deliveryAddress?: string | null;
    deliverySchedule?: string | null;
    formattedDeliverySchedule?: string | null;
    lastStatusUpdate?: string | null;
    formattedLastStatusUpdate?: string | null;
    notes?: string | null;
    internalNotes?: string | null;
    specialInstructions?: string[] | null;
    customer: Customer;
    commissionLogs: ReferralLog[];
    employee: Employee;
    orderItems: OrderItem[];
    orderItemsCount?: number;
    paymentMethod?: OrderPaymentMethod | null;
    orderStatusHistories?: OrderStatusHistory[];
    orderPaymentLogs?: OrderPaymentLog[];
    orderPaymentLogsCount?: number;
    canPay?: boolean;
    canScheduleDelivery: boolean;
    requiresPaymentBeforeDelivery: boolean;
    review?: OrderReview | null;
    hasReview: boolean;
    outlet?: Outlet | null;
    customerAddress?: CustomerAddress | null;
    createdAt: string;
    updatedAt: string;
    formattedCreatedAt: string;
    formattedUpdatedAt: string;
    deletedAt?: string | null;
}

/**
 * Order filter interface
 */
export interface OrderFilters extends BaseFilters {
    estimatedCompletionFrom: any;
    estimatedCompletionTo: any;
    status?: string;
    paymentStatus?: string;
    outletId?: number;
    customerId?: number;
    employeeId?: number;
    orderDateFrom?: string;
    orderDateTo?: string;
    totalAmountMin?: number;
    totalAmountMax?: number;
}

/**
 * Order sort options
 */
export interface OrderSortOptions extends BaseSortOptions {
    column:
        | "orderNumber"
        | "status"
        | "paymentStatus"
        | "totalAmount"
        | "paidAmount"
        | "remainingAmount"
        | "orderDate"
        | "estimatedCompletion"
        | "actualCompletion"
        | "createdAt"
        | "updatedAt";
}

/**
 * Order form data
 */
export interface OrderFormData {
    customerId: number;
    employeeId?: number;
    notes?: string;
    internalNotes?: string;
    specialInstructions?: string[];
    estimatedCompletion?: string;
    orderItems: OrderItemFormData[];
}

/**
 * Order item form data
 */
export interface OrderItemFormData {
    laundryServiceId: number;
    quantity: number;
    itemNotes?: string;
}

File: laundry/wash_wallet/packages/wash_wallet_domain/lib/src/entities/order.dart
import 'package:equatable/equatable.dart';
import 'customer.dart';
import 'employee.dart';
import 'outlet.dart';
import 'order_item.dart';

class Order extends Equatable {
  final int id;
  final String orderNumber;
  final String status;
  final String? statusLabel;
  final String? statusBadgeVariant;
  final int completionPercentage;
  final String paymentStatus;
  final String? paymentStatusLabel;
  final String? paymentStatusBadgeVariant;
  final String? deliveryType;
  final String? deliveryTypeLabel;
  final bool canPay;
  final bool canScheduleDelivery;
  final bool requiresPaymentBeforeDelivery;
  final String? source;
  final String? sourceLabel;
  final String? paymentMethod;
  final int customerId;
  final int? customerAccountId;
  final int employeeId;
  final int outletId;
  final int? customerAddressId;
  final int? updatedBy;
  final Customer? customer;
  final dynamic customerAccount;
  final Employee? employee;
  final Outlet? outlet;
  final dynamic customerAddress;
  final Employee? statusUpdater;
  final List<dynamic> commissionLogs;
  final dynamic review; // OrderReview di customer app
  final bool hasReview;
  final double subtotal;
  final double pickupFee;
  final String? formattedPickupFee;
  final double deliveryFee;
  final String? formattedDeliveryFee;
  final double discountAmount;
  final double taxAmount;
  final double totalAmount;
  final double paidAmount;
  final double remainingAmount;
  final String? formattedSubtotal;
  final String? formattedDiscountAmount;
  final String? formattedTaxAmount;
  final String? formattedTotalAmount;
  final String? formattedPaidAmount;
  final String? formattedRemainingAmount;
  final String? midtransOrderId;
  final String? midtransTransactionId;
  final String? qrUrl;
  final DateTime? orderDate;
  final String? formattedOrderDate;
  final DateTime? estimatedCompletion;
  final String? formattedEstimatedCompletion;
  final DateTime? actualCompletion;
  final String? formattedActualCompletion;
  final DateTime? pickupDate;
  final String? formattedPickupDate;
  final DateTime? deliveryDate;
  final String? formattedDeliveryDate;
  final DateTime? deliverySchedule;
  final String? formattedDeliverySchedule;
  final String? deliveryAddress;
  final DateTime? lastStatusUpdate;
  final String? formattedLastStatusUpdate;
  final String? pickupType;
  final String? pickupAddress;
  final DateTime? pickupSchedule;
  final String? formattedPickupSchedule;
  final String? notes;
  final String? internalNotes;
  final dynamic specialInstructions;
  final List<OrderItem>? orderItems;
  final int orderItemsCount;
  final DateTime? createdAt;
  final DateTime? updatedAt;
  final String? formattedCreatedAt;
  final String? formattedUpdatedAt;
  final DateTime? deletedAt;

  const Order({
    required this.id,
    required this.orderNumber,
    required this.status,
    this.statusLabel,
    this.statusBadgeVariant,
    this.completionPercentage = 0,
    required this.paymentStatus,
    this.paymentStatusLabel,
    this.paymentStatusBadgeVariant,
    this.deliveryType,
    this.deliveryTypeLabel,
    this.canPay = false,
    this.canScheduleDelivery = false,
    this.requiresPaymentBeforeDelivery = false,
    this.source,
    this.sourceLabel,
    this.paymentMethod,
    required this.customerId,
    this.customerAccountId,
    required this.employeeId,
    required this.outletId,
    this.customerAddressId,
    this.updatedBy,
    this.customer,
    this.customerAccount,
    this.employee,
    this.outlet,
    this.customerAddress,
    this.statusUpdater,
    this.commissionLogs = const [],
    this.review,
    this.hasReview = false,
    this.subtotal = 0,
    this.pickupFee = 0,
    this.formattedPickupFee,
    this.deliveryFee = 0,
    this.formattedDeliveryFee,
    this.discountAmount = 0,
    this.taxAmount = 0,
    this.totalAmount = 0,
    this.paidAmount = 0,
    this.remainingAmount = 0,
    this.formattedSubtotal,
    this.formattedDiscountAmount,
    this.formattedTaxAmount,
    this.formattedTotalAmount,
    this.formattedPaidAmount,
    this.formattedRemainingAmount,
    this.midtransOrderId,
    this.midtransTransactionId,
    this.qrUrl,
    this.orderDate,
    this.formattedOrderDate,
    this.estimatedCompletion,
    this.formattedEstimatedCompletion,
    this.actualCompletion,
    this.formattedActualCompletion,
    this.pickupDate,
    this.formattedPickupDate,
    this.deliveryDate,
    this.formattedDeliveryDate,
    this.deliverySchedule,
    this.formattedDeliverySchedule,
    this.deliveryAddress,
    this.lastStatusUpdate,
    this.formattedLastStatusUpdate,
    this.pickupType,
    this.pickupAddress,
    this.pickupSchedule,
    this.formattedPickupSchedule,
    this.notes,
    this.internalNotes,
    this.specialInstructions,
    this.orderItems,
    this.orderItemsCount = 0,
    this.createdAt,
    this.updatedAt,
    this.formattedCreatedAt,
    this.formattedUpdatedAt,
    this.deletedAt,
  });

  factory Order.fromModel(dynamic model) {
    return Order(
      id: model.id,
      orderNumber: model.orderNumber,
      status: model.status,
      statusLabel: model.statusLabel,
      statusBadgeVariant: model.statusBadgeVariant,
      completionPercentage: model.completionPercentage,
      paymentStatus: model.paymentStatus,
      paymentStatusLabel: model.paymentStatusLabel,
      paymentStatusBadgeVariant: model.paymentStatusBadgeVariant,
      deliveryType: model.deliveryType,
      deliveryTypeLabel: model.deliveryTypeLabel,
      canPay: model.canPay ?? false,
      canScheduleDelivery: model.canScheduleDelivery ?? false,
      requiresPaymentBeforeDelivery: model.requiresPaymentBeforeDelivery ?? false,
      source: model.source,
      sourceLabel: model.sourceLabel,
      paymentMethod: model.paymentMethod,
      customerId: model.customerId,
      customerAccountId: model.customerAccountId,
      employeeId: model.employeeId,
      outletId: model.outletId,
      customerAddressId: model.customerAddressId,
      updatedBy: model.updatedBy,
      customer: model.customer?.toEntity(),
      customerAccount: model.customerAccount?.toEntity(),
      employee: model.employee?.toEntity(),
      outlet: model.outlet?.toEntity(),
      customerAddress: model.customerAddress,
      statusUpdater: model.statusUpdater?.toEntity(),
      commissionLogs: model.commissionLogs ?? const [],
      subtotal: model.subtotal,
      pickupFee: model.pickupFee,
      formattedPickupFee: model.formattedPickupFee,
      deliveryFee: model.deliveryFee,
      formattedDeliveryFee: model.formattedDeliveryFee,
      discountAmount: model.discountAmount,
      taxAmount: model.taxAmount,
      totalAmount: model.totalAmount,
      paidAmount: model.paidAmount,
      remainingAmount: model.remainingAmount,
      formattedSubtotal: model.formattedSubtotal,
      formattedDiscountAmount: model.formattedDiscountAmount,
      formattedTaxAmount: model.formattedTaxAmount,
      formattedTotalAmount: model.formattedTotalAmount,
      formattedPaidAmount: model.formattedPaidAmount,
      formattedRemainingAmount: model.formattedRemainingAmount,
      midtransOrderId: model.midtransOrderId,
      midtransTransactionId: model.midtransTransactionId,
      qrUrl: model.qrUrl,
      orderDate: model.orderDate,
      formattedOrderDate: model.formattedOrderDate,
      estimatedCompletion: model.estimatedCompletion,
      formattedEstimatedCompletion: model.formattedEstimatedCompletion,
      actualCompletion: model.actualCompletion,
      formattedActualCompletion: model.formattedActualCompletion,
      pickupDate: model.pickupDate,
      formattedPickupDate: model.formattedPickupDate,
      deliveryDate: model.deliveryDate,
      formattedDeliveryDate: model.formattedDeliveryDate,
      deliverySchedule: model.deliverySchedule,
      formattedDeliverySchedule: model.formattedDeliverySchedule,
      deliveryAddress: model.deliveryAddress,
      lastStatusUpdate: model.lastStatusUpdate,
      formattedLastStatusUpdate: model.formattedLastStatusUpdate,
      pickupType: model.pickupType,
      pickupAddress: model.pickupAddress,
      pickupSchedule: model.pickupSchedule,
      formattedPickupSchedule: model.formattedPickupSchedule,
      notes: model.notes,
      internalNotes: model.internalNotes,
      specialInstructions: model.specialInstructions,
      orderItems: model.orderItems?.map((e) => OrderItem.fromModel(e)).toList(),
      orderItemsCount: model.orderItemsCount,
      createdAt: model.createdAt,
      updatedAt: model.updatedAt,
      formattedCreatedAt: model.formattedCreatedAt,
      formattedUpdatedAt: model.formattedUpdatedAt,
      deletedAt: model.deletedAt,
    );
  }

  @override
  List<Object?> get props => [
    id,
    orderNumber,
    status,
    statusLabel,
    statusBadgeVariant,
    completionPercentage,
    paymentStatus,
    paymentStatusLabel,
    paymentStatusBadgeVariant,
    deliveryType,
    deliveryTypeLabel,
    canPay,
    canScheduleDelivery,
    requiresPaymentBeforeDelivery,
    source,
    sourceLabel,
    paymentMethod,
    customerId,
    customerAccountId,
    employeeId,
    outletId,
    customerAddressId,
    updatedBy,
    customer,
    customerAccount,
    employee,
    outlet,
    customerAddress,
    statusUpdater,
    commissionLogs,
    review,
    hasReview,
    subtotal,
    pickupFee,
    formattedPickupFee,
    deliveryFee,
    formattedDeliveryFee,
    discountAmount,
    taxAmount,
    totalAmount,
    paidAmount,
    remainingAmount,
    formattedSubtotal,
    formattedDiscountAmount,
    formattedTaxAmount,
    formattedTotalAmount,
    formattedPaidAmount,
    formattedRemainingAmount,
    midtransOrderId,
    midtransTransactionId,
    qrUrl,
    orderDate,
    formattedOrderDate,
    estimatedCompletion,
    formattedEstimatedCompletion,
    actualCompletion,
    formattedActualCompletion,
    pickupDate,
    formattedPickupDate,
    deliveryDate,
    formattedDeliveryDate,
    deliverySchedule,
    formattedDeliverySchedule,
    deliveryAddress,
    lastStatusUpdate,
    formattedLastStatusUpdate,
    pickupType,
    pickupAddress,
    pickupSchedule,
    formattedPickupSchedule,
    notes,
    internalNotes,
    specialInstructions,
    orderItems,
    orderItemsCount,
    createdAt,
    updatedAt,
    formattedCreatedAt,
    formattedUpdatedAt,
    deletedAt,
  ];
}
